src/extensions/default/PhpTooling/main.js

```diff
@@ -0,0 +1,244 @@
+        ExtensionUtils = brackets.getModule("utils/ExtensionUtils"),
+        ProjectManager = brackets.getModule("project/ProjectManager"),
+        EditorManager =  brackets.getModule("editor/EditorManager"),
+        LanguageManager =  brackets.getModule("language/LanguageManager"),
+        CodeHintManager = brackets.getModule("editor/CodeHintManager"),
+        ParameterHintManager = brackets.getModule("features/ParameterHintsManager"),
+        JumpToDefManager = brackets.getModule("features/JumpToDefManager"),
+        CodeInspection = brackets.getModule("language/CodeInspection"),
+        DefaultProviders = brackets.getModule("languageTools/DefaultProviders"),
+        CodeHintsProvider = require("CodeHintsProvider").CodeHintsProvider,
+        DefaultEventHandlers = brackets.getModule("languageTools/DefaultEventHandlers"),
+        PreferencesManager  = brackets.getModule("preferences/PreferencesManager"),
+        Strings             = brackets.getModule("strings"),
+        Dialogs             = brackets.getModule("widgets/Dialogs"),
+        DefaultDialogs      = brackets.getModule("widgets/DefaultDialogs"),
+        Commands               = brackets.getModule("command/Commands"),
+        CommandManager         = brackets.getModule("command/CommandManager"),
+        StringUtils             = brackets.getModule("utils/StringUtils");
+
+    var clientFilePath = ExtensionUtils.getModulePath(module, "client.js"),
+        clientName = "PhpClient",
+        _client = null,
+        phpConfig = {
+            enablePhpTooling: true,
+            executablePath: "php",
+            memoryLimit: "4095M",
+            validateOnSave: true
+        },
+        DEBUG_OPEN_PREFERENCES_IN_SPLIT_VIEW  = "debug.openPrefsInSplitView",
+        phpServerRunning = false,
+        serverCapabilities,
+        currentRootPath;
+
+    PreferencesManager.definePreference("php", "object", phpConfig, {
+        description: Strings.DESCRIPTION_PHP_TOOLING_CONFIGURATION
+    });
+
+    PreferencesManager.on("change", "php", function () {
+        var newPhpConfig = PreferencesManager.get("php");
+
+        if ((newPhpConfig["executablePath"] !== phpConfig["executablePath"])
+                || (newPhpConfig["enablePhpTooling"] !== phpConfig["enablePhpTooling"])) {
+            phpConfig = newPhpConfig;
+            runPhpServer();
+            return;
+        }
+        phpConfig = newPhpConfig;
+    });
+
+    var handleProjectOpen = function (event, directory) {
+        _client.stop()
+            .done(function () {
+                setTimeout(function () {
+                    _client.start({
+                        rootPath: directory.fullPath
+                    });
+                }, 1500);
+            }).fail(function () {
+                console.log("Error encountered while stoping Php Server.");
+            });
+    };
+
+    function registerToolingProviders() {
+        var chProvider = new CodeHintsProvider(_client),
```

Comment:
do this instead:
```
var chProvider = new DefaultProviders.CodeHintsProvider(_client),
chProvider.getHints = yourOwnGetHintsFunction;
```

Reply:
I don't see this gives us any benefits, I understand that I have changes only in getHints function but still I have to rewrite all Util functions which are defined in default code Hints provider and I have to access "this" object, If i use "this" variable in normal functions that will look weird. this seems to be an hack code not recommended coding practise. 
If you are going to externalise all util functions, then I can reuse your util functions from default provider.